商品購入機能

aiueo818 opened the pull request:
# What
商品購入機能の実装
# Why
最終課題において順番に課題に取り組むため

# 動画、写真
必要な情報を適切に入力して「購入」ボタンを押すと、商品の購入ができる動画
https://gyazo.com/b4f2e4e432862b1416b7c819f00384bd
入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示される動画
https://gyazo.com/f1b02f6284b01e4982fa018d77f70881
 ログイン状態の場合でも、URLを直接入力して自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
https://gyazo.com/eaac6ed288d620a7b6acd8e6e0367691
ログイン状態の場合でも、URLを直接入力して自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移する動画
https://gyazo.com/fa0d6e02516147bc5c291f036190e223
ログアウト状態の場合は、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画
https://gyazo.com/025c85bdbf92ca73712bd140aa0b8b7a
売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品一覧、商品詳細機能実装時に未実装であった場合）
https://gyazo.com/335331205ad2f921efa172b044552b52
ログイン状態の場合でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されない動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/b73274c18223a8fdc44c99da0ab9153b
ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
https://gyazo.com/ac7dd13339577d51a1e5be43eb6f66aa
テスト結果の画像
https://gyazo.com/189a6fea048ab9d4a59f7ab77f45b146

aiueo818 committed:
商品購入機能

# What
商品購入機能の実装
# Why
最終課題において順番に課題に取り組むため

div-reviewer commented on app/controllers/buys_controller.rb:
```diff
@@ -0,0 +1,44 @@
+  end
+
+  def create
+    @item = Item.find(params[:item_id])
```
```ruby  
  @item = Item.find(params[:item_id])
```
こちらの記述が複数のメソッドで使用されているかと思われます。
よってprivate以下にメソッド化し、before_actionで呼び出しましょう。

**理由**
可読性を上げるためです。
複数回使用する変数をメソッド化することによってまず可読性が上がります。また保守（修作業など）が発生した際も手を加えやすくなります。

```ruby
【実装イメージ】
before_action :set_item, ~~~

def set_item
 @item = Item.find(params[:item_id])
end
```

>参考カリキュラム
フェーズ：応用
タイトル：Ruby on Rails（PicTweet）〜ツイートの詳細を表示しよう〜
目次：同じコードをメソッドにまとめよう

div-reviewer commented on app/controllers/buys_controller.rb:
```diff
@@ -0,0 +1,44 @@
+    @buy_address = BuyAddress.new(buy_params)
+    price = @item.price
+    if @buy_address.valid?
+      pay_item(price)
```
```ruby
 def pay_item(price)
    Payjp.api_key = ENV["PAYJP_SECRET_KEY"]
      Payjp::Charge.create(
        amount: price,
↓
 def pay_item
    Payjp.api_key = ENV["PAYJP_SECRET_KEY"]
      Payjp::Charge.create(
        amount: @item.price,
```

今回createアクション内で`@item（商品）`をすでに作成されているため、再代入ではなく、`@item`を使用した形にしましょう。そうすることで以下2行削除することが可能です。

```ruby
   price = @item.price
      pay_item(price)
```

div-reviewer commented on app/controllers/buys_controller.rb:
```diff
@@ -0,0 +1,44 @@
+  end
+
+  def show
+  end
```
今回購入機能の詳細ページはないため、削除しましょう。

div-reviewer commented on app/controllers/buys_controller.rb:
```diff
@@ -0,0 +1,44 @@
+    @item = Item.find(params[:item_id])
+    @buy_address = BuyAddress.new
+    
+    if @item.buy != nil
+      redirect_to root_path
+    end
```
以下がまだ未実装のようです。条件を確認し、実装しましょう！

```ruby
- ログイン状態の出品者が、URLを直接入力して自身の出品した商品購入ページに遷移しようとすると、トップページに遷移すること
```

div-reviewer commented on app/models/buy_address.rb:
```diff
@@ -0,0 +1,20 @@
+  include ActiveModel::Model
+  attr_accessor :user, :item_id, :post_code, :shipping_id, :city, :street_address, :building, :phone_number, :buy, :user_id, :token
+
+  validates :shipping_id, :city, :street_address, :item_id, :token, presence: true
```
presence: true（空で保存してほしくないという制約）の中に「user_id」も追加しましょう。

**理由**
普通のモデルファイルであれば、belongs_toをつけると自動的に「必ず存在しなければならない」というvalidationが付きます。
しかし、今回はフォームオブジェクトを使用しているためアソシエーションで制限を課せない為、改めてpresence: trueの中に記述する必要があります。

購入管理のテーブルでは、userのidが必ず存在する必要があり、validationをかけることで、よりアプリケーションの実装内容を保護することができます。

div-reviewer replied:
郵便番号や電話番号でwith_optionsを使用してくださっているので、以下の内容でもwith_optionsを使ってみましょう。

```ruby
validates :shipping_id, :city, :street_address, :item_id, :token, presence: true
```

div-reviewer commented on app/views/items/show.html.erb:
```diff
@@ -34,7 +36,7 @@
         <%# 商品が売れていない場合はこちらを表示しましょう %>
-    <%= link_to "購入画面に進む", "item_path" ,class:"item-red-btn"%>
+    <%= link_to "購入画面に進む", item_buys_path(@item.id) ,class:"item-red-btn" %>
     <%# //商品が売れていない場合はこちらを表示しましょう %>
```
```ruby
<%# //商品が売れていない場合はこちらを表示しましょう %>
<%# 商品が売れていない場合はこちらを表示しましょう %>
```

実装`済みの`案内用文は削除しましょう。

**理由**
コードレビューは一旦完成したコードに対して行うものであるのと、企業様に提出する際に実装案内のコメントが残っていることは好ましくないためです。
もし学習メモとして必要であれば、第三者でもわかるように「残す目的」「なぜ必要か」分かるよう、`「学習メモ」`等分かりやすいコメントとして書いておくことをおすすめいたします。

■補足
他にも不要な案内文があれば、削除しましょう。

div-reviewer commented on spec/factories/buy_addresses.rb:
```diff
@@ -0,0 +1,12 @@
+    street_address { '1-1' }
+    building { '東京ハイツ' }
+    phone_number { '09012341234' }
+    item_id { 5 }
```
現在問題なく動いているかと思いますが、商品_idデータの作成方法について修正しましょう。

**理由**
外部キーに直接数字を入れてしまうと本当にそのレコードが存在するか分からないため、データの信頼性が低くなってしまいます。
実際に存在するレコードのidを入れたい為、Factorybotでレコードを生成し、フォームオブジェクトのインスタンスに組み込みましょう。

**改善案**
```ruby
例)
before do
①ユーザー = Factorybotを用いてデータを生成（DBに保存）
②商品 = Factorybotを用いてデータを生成（DBに保存）
③@フォームオブジェクト = FactoryBot.build(:フォームオブジェクト, ユーザー_id: ユーザー.id , 商品_id: 商品.id)
end
```

■参考
```ruby
コース：最終課題
タイトル：最終課題補足カリキュラム〜1つのフォームから複数のテーブルに情報を保存するアプリを作成しよう〜

テストコードを書こう
```

**mysqlエラーが出た時の対処**
今回beforeアクション内にデータが多い為、Mysqlエラーが出てしまう可能性があります。エラーが出た場合は以下カリキュラムを参考に解消しましょう。

```ruby
コース：応用
タイトル：Ruby on Rails（ChatApp）〜ChatAppのモデル単体テストコードを書こう〜

「Messageモデルをテストしよう」の項目
```

koyamatakumi replied:
```ruby
 item_id { 5 }
```
こちらの記述は不要となるため消しておきましょう。

div-reviewer commented on app/models/buy_address.rb:
```diff
@@ -0,0 +1,20 @@
+
+  with_options presence: true, format: { with: /\A\d{10,11}\z/, message: 'はハイフン無しで入力してください' } do
+    validates :phone_number
+  end  
```
商品出品同様、購入機能でも都道府県はジャンルの選択が「---」の時は保存できないようにするバリデーションを設けましょう。

div-reviewer commented on spec/models/buy_address_spec.rb:
```diff
@@ -0,0 +1,65 @@
+        @buy_address.phone_number = '-'
+        @buy_address.valid?
+        expect(@buy_address.errors.full_messages).to include("Phone number はハイフン無しで入力してください")
+      end
```
下記のテスト検証を追加しましょう！

```ruby
異常系
■都道府県
・都道府県に「---」が選択されている場合は購入できない
■電話番号 
・電話番号が9桁以下では購入できない
・電話番号が12桁以上では購入できない
■購入者情報
・userが紐付いていなければ購入できない
■商品情報
・itemが紐付いていなければ購入できない
```

div-reviewer reviewed (commented):
aiueo818様、学習お疲れ様です！
商品購入の初回レビュー依頼ありがとうございます。プルリクエスト確認させていただきました。

完成度高く実装されているかと思われます。今までの学習がバッチリ成果として出ていますね。大変素晴らしいです。

修正点をコメントさせて頂きましたので、ご確認をお願い致します。

【次回プルリクエスト提出する際の注意点】
```ruby
①修正が終わりましたら、`コミットとプッシュ`をして、`プルリクエストに修正内容が反映されているか確認`しましょう。
②既に頂いている挙動確認の画像や動画が修正範囲だった場合、最新の画像や動画の提出をお願い致します。
③googleフォームの送信をお願いいたします。
```

**ご不明点や疑問点ございましたら、チャット・コネクトメンターにご相談ください！
どうぞよろしくおねがいします。**

aiueo818 committed:
商品購入機能の修正

# What
レビューいただいた箇所の修正
# Why
商品購入機能を正しく実装するため

aiueo818 commented:
# 動画、写真（更新したもののみ）
ログイン状態の出品者が、URLを直接入力して自身の出品した商品購入ページに遷移しようとすると、トップページに遷移すること
https://gyazo.com/2efa877a8e54d3a7cdbe131d53cabb4a
テストに成功した写真
https://gyazo.com/3890af146a65395da32c3849200346f7

koyamatakumi commented on app/controllers/buys_controller.rb:
```diff
@@ -0,0 +1,47 @@
+class BuysController < ApplicationController
+  before_action :set_item, only: [:index, :create]
+
+  def index
```
ログアウト中のユーザーが購入ページへリクエストを送った場合、ログインページへ遷移させるようにしましょう。

koyamatakumi replied:
https://github.com/aiueo818/furima-40136/pull/19/files#r1447044890
上記コメントの変更が確認できませんでした。
:authenticate_user!メソッドを用いることができるかと存じます。改めてご確認の上修正をお願いいたします。

koyamatakumi commented on app/controllers/buys_controller.rb:
```diff
@@ -0,0 +1,47 @@
+  private
+
+  def buy_params
+    params.require(:buy_address).permit(:post_code, :shipping_id, :city, :street_address, :building, :phone_number, :buy).merge(user_id: current_user.id, item_id: params[:item_id], token: params[:token])
```
```ruby    
params.require(:buy_address).permit(:post_code, :shipping_id, :city, :street_address, :building, :phone_number, :buy).merge(user_id: current_user.id, item_id: params[:item_id], token: params[:token])
```
permitに不要な引数が1つございます。ご確認の上削除しましょう。

koyamatakumi commented on app/models/buy_address.rb:
```diff
@@ -0,0 +1,25 @@
+class BuyAddress
+  include ActiveModel::Model
+  attr_accessor :user, :item_id, :post_code, :shipping_id, :city, :street_address, :building, :phone_number, :buy, :user_id, :token
```
```ruby
  attr_accessor :user, :item_id, :post_code, :shipping_id, :city, :street_address, :building, :phone_number, :buy, :user_id, :token
```
:user、:buyは不要なため消しておきましょう。

koyamatakumi reviewed (commented):
aiueo818様 
レビュー依頼フォームの送信ありがとうございます。 修正いただいた内容を確認させていただき、追加でコメントいたしましたのでご確認お願いいたします。 

修正後はコミット、プッシュを忘れずに行い、改めてレビュー依頼フォームの送信をお願い致します。着実にゴールに近づいています。ラストスパート頑張っていきましょう！

aiueo818 committed:
商品購入機能の修正２

# What
レビューいただいた箇所の修正
# Why
商品購入機能を正しく実装するため

aiueo818 commented:
# 動画（更新したもののみ）
ログアウト中のユーザーが購入ページへリクエストを送った場合、ログインページへ遷移させる動画
https://gyazo.com/d6e90bcc2dd8ca51b100b9a2d3bb4a7b

koyamatakumi reviewed (commented):
aiueo818様 
レビュー依頼フォームの送信ありがとうございます。 修正いただいた内容を確認させていただき、追加でコメントいたしましたのでご確認お願いいたします。 

修正後はコミット、プッシュを忘れずに行い、改めてレビュー依頼フォームの送信をお願い致します。着実にゴールに近づいています。ラストスパート頑張っていきましょう！

aiueo818 committed:
商品購入機能の修正３

# What
レビューいただいた箇所で修正していない部分があったため、そちらの修正を行いました。大変申し訳ありません。
# Why
商品購入機能を正しく実装するため

aiueo818 commented:
# 動画（修正点の部分のみ）
ログアウト中のユーザーが購入ページへリクエストを送った場合、ログインページへ遷移させる動画
https://gyazo.com/e2257918e6b388e7e0ee288ef72f0a65

yuri240 reviewed (commented):
修正とレビュー依頼ありがとうございます。

しっかりと実装できていることが確認できましたので、
**LGTM**とさせていただきます！

こちらのブランチは忘れずにMergeを行っていただき、
ローカルのMasterにPullしておきましょう。

ここまでの実装、大変お疲れ様でした。
引き続き、応援しております。

aiueo818 commented:
ありがとうございます！！！！
